src/tunnel.cc

```diff
@@ -672,15 +781,40 @@ tunnelTimeout(const CommTimeoutCbParams &io)
     /* Temporary lock to protect our own feets (comm_close -> tunnelClientClosed -> Free) */
     CbcPointer<TunnelStateData> safetyLock(tunnelState);
 
-    tunnelState->client.closeIfOpen();
-    tunnelState->server.closeIfOpen();
+    tunnelState->closeConnections();
 }
 
 void
-TunnelStateData::Connection::closeIfOpen()
+TunnelStateData::closeConnections()
 {
-    if (Comm::IsConnOpen(conn))
+    if (Comm::IsConnOpen(server.conn))
+        server.conn->close();
+    if (Comm::IsConnOpen(client.conn))
+        client.conn->close();
+}
+
+void
+TunnelStateData::Connection::closeQuietly()
+{
+    assert(!writer);
+
+    if (Comm::IsConnOpen(conn)) {
+        Must(closer);
+        comm_remove_close_handler(conn->fd, closer);
+        closer = nullptr;
         conn->close();
+        conn = nullptr;
+        return;
+    }
+
+    if (closer) {
+        Must(conn);
+        Must(fd_table[conn->fd].closing()); // explains false IsConnOpen()
```

Comment:
This code looks problematic to me:: both true conditions !Comm::IsConnOpen(conn) and conn!=null imply that fd==-1, but we cannot use this as an array index.

Reply:
`conn && !IsConnOpen(conn)` does not imply `fd==-1` because Connection::isOpen() tests (among other things) `fd_table[fd].flags.open`.

However, I agree that we should add another Must() to make sure that conn->fd is set. There may be similar problems in some other fde::closing() calls (inside and/or outside this PR). Please check and fix (the ones inside this PR).

Reply:
Here are these checks:

```
Comm::IsConnOpen(const Comm::ConnectionPointer &conn) { return conn != NULL && conn->isOpen(); }

bool Connection::isOpen() const { return (fd >= 0); }

```

I do not see here fd_table[fd].flags.open check.

Reply:
@eduard-bagdasaryan, you are right. I must have looked at isOpen() instead of IsConnOpen(). The different meaning of these functions is probably the source of some old bugs, but we are not going to address that in this PR, of course. I am glad you spotted this new bug! I hope branch commit b722bc4 addresses your concern. Please check and adjust as needed.

> @yadij: In its current placement this if-statement block should not be doing anything related to conn, only cancel and discard the closer.

And it should also clear the `conn` because the tunnel is completely done working with that connection now. The `conn` member is not nil inside the `if (closer)` statement (when Comm::IsConnOpen() is false), because when the tunnel calls conn->close(), it does not clear the `conn` member. This `conn` clearing is probably not critical because we usually check Comm::IsConnOpen() instead of checking the `conn` pointer itself, but this clearing is the right action here.

When the closeQuietly() method returns, the affected tunnel side should no longer be associated with any Connection object, and no Connection callbacks should reach the tunnel. Please keep in mind that this method is not a connection closing callback. It is a method designed to get rid of the connection (closing it if needed) without allowing the previously assigned closing callbacks (if any) to reach the tunnel.

Reply:
I think your b722bc4 solution fixed the problem.

Reply:
> Please keep in mind that this method is not a connection closing callback.

... yet it is called "close", documented as closing the connection, and calls close at every opportunity.

> It is a method designed to get rid of the connection (closing it if needed) without allowing the previously assigned closing callbacks (if any) to reach the tunnel.

The design of Comm::Connection is that code with that requirements should cancel their closer callback and set their Pointer to nullptr. Not call close(). Anything sharing the conn object will then be free to continue using it, and if there are no sharing the conn will close itself cleanly.

There is one oddity with tunnels *if* the connection has possibly been used (dirty?) or the client was told it got opened (200 status message) then is *also* has to be closed for safety. That is a feature/promise of the tunnel mechanism - *not* the Comm::Connection API.

IMO this method should look like this:
``
{
    assert(!writer);

    if (closer) {
        // cancel in case it is already scheduled
        closer->cancel("no longer needed by tunnel");
    }

    if (Comm::IsConnOpen(conn)) {
        Must(closer);
        comm_remove_close_handler(conn->fd, closer);
    }

    conn->close();
    conn = nullptr;
    closer = nullptr;
}
``

Reply:
Maybe with a check for conn being nullptr at the end there if that case can ever happen.

Reply:
> Alex: It is a method designed to get rid of the connection (closing it if needed) without allowing the previously assigned closing callbacks (if any) to reach the tunnel.

> Amos: The design of Comm::Connection is that code with that requirements should cancel their closer callback and set their Pointer to nullptr. Not call close().

@yadij, the first 2-3 paragraphs of your [comment](https://github.com/squid-cache/squid/pull/489#discussion_r376959427) reflect some kind of miscommunication -- I think we have been talking about different code areas and/or different method requirements while thinking that we are discussing the same thing. I suggest exiting that part of the discussion with the following basic rule of thumb:

* When an owner of a Comm::Connection object decides that it is time to close the underlying transport connection, that owner should call Comm::Connection::close().

Both the PR code and your suggestion (quoted below in its polished version, per your [comment](https://github.com/squid-cache/squid/pull/489#discussion_r376961468)) comply with the above rule.

----

> Amos: IMO this method should look like this:


```C++
if (closer) {
    // cancel in case it is already scheduled
    closer->cancel("no longer needed by tunnel");
}

if (Comm::IsConnOpen(conn)) {
    Must(closer);
    comm_remove_close_handler(conn->fd, closer);
}

if (conn) {
    conn->close();
    conn = nullptr;
}

closer = nullptr;
```

AFAICT, both the PR code and the above suggestion accomplish the same thing. Your suggestion focuses on "clearing" or "getting rid of" seemingly independent connection artifacts -- closer, closer, and conn (and closer again). Those objects are actually dependent on each other and their dependencies are important for understanding and writing correct code. The code also has small problems (discussed/fixed later in this comment).

Overall, I prefer the PR version because it focuses on separating the two important but not obvious use cases (the first two bullets below), simplifying code logic by removing several hidden/dangerous dependencies or misleading if statements. It also allows us to add useful assertions. The third use case is trivial, but we should add a comment to document it _explicitly_, like we already do with the other two cases. All three use cases are mutually exclusive and all focus on the same set of {TunnelStateData,Comm}::Connection objects:

* Connection-owning TunnelStateData did not call conn->close() before this closeQuietly() call.
* Connection-owning TunnelStateData called conn->close() before this closeQuietly() call.
* TunnelStateData does not own the connection (at the time of this closeQuietly() call).


There are also several small problems with the suggested code quoted above:

1. The code manually cancels the callback and then calls comm_remove_close_handler() which also (naturally) cancels the same callback. Ideally, such repeated cancellations should be avoided.
2. The code introduces a new state -- `closer` is still set but the callback has already been canceled. Yes, this state is short-lived, but, ideally, the `closer` pointer should be cleared immediately after we get rid of the callback, reducing the chances of future separation bugs, removing surprises during exception-triggered cleanup, and simplifying object state. There should be no third state -- either we are asking Comm to call us back or we are not.
3. The code implies that IsConnOpen() might be true even without `closer`. That implication is wrong -- we should never have an open connection without a closing callback! It may seem like a minor complication -- it _is_ possible to get rid of each object independently -- but the tunneling code will become more complex if it does not maintain/rely on that simpler state. This is similar to adding an innocent-looking `if (foo)` statement where `foo` cannot be nil: Sooner or later, the code will start partially handing nil foos, nil foos might even become possible (breaking legacy code), and, in a few years, it would become very difficult to untangle that mess. The rule of thumb here is that a connection owner must maintain a closing callback -- the two objects are tied together. (Eventually, we will add APIs to better express connection ownership)
4. The code implies that conn might be true even without `closer`. Same reasoning as in item 3.


The above problems can be fixed, distilling the essential differences between PR and the suggested code. For example:

Item 1 can be addressed by avoiding the split of the callback cancellation code, moving it all inside the dedicated `closer` check:

```C++
if (closer) {
    if (Comm::IsConnOpen(conn))
        comm_remove_close_handler(conn->fd, closer);
    else
        closer->cancel("no longer needed by tunnel");
}

if (conn) {
    conn->close();
    conn = nullptr;
}

closer = nullptr;
```

Note that the above essentially fixes item 3 as well, which we can now express by adding a Must():

```C++
if (closer) {
    if (Comm::IsConnOpen(conn))
        comm_remove_close_handler(conn->fd, closer);
    else
        closer->cancel("no longer needed by tunnel");
} else {
    // open connection requires a closer
    Must(!Comm::IsConnOpen(conn));
}

if (conn) {
    conn->close();
    conn = nullptr;
}

closer = nullptr;
```

Similarly, we can fix item 4 by moving the close() call and the `conn` nullification to where `conn` is present. The Must() we just added above can be strengthened accordingly:

```C++
if (closer) {
    Must(conn);
    if (Comm::IsConnOpen(conn)) {
        comm_remove_close_handler(conn->fd, closer);
        conn->close();
        conn = nullptr;
    } else {
        closer->cancel("no longer needed by tunnel");
        conn = nullptr;
    }
} else {
    Must(!conn); // connection requires a closer
}

closer = nullptr;
```

Now we can address item 2 by nullifying `closer` ASAP:

```C++
if (closer) {
    if (Comm::IsConnOpen(conn)) {
        comm_remove_close_handler(conn->fd, closer);
        closer = nullptr;
        conn->close();
        conn = nullptr;
    } else {
        closer->cancel("no longer needed by tunnel");
        closer = nullptr;
        conn = nullptr;
    }
} else {
    Must(!conn); // connection requires a closer
}
```

To simplify the comparison with the PR code, let's convert the above to use the if-return style:

```C++
if (closer) {
    if (Comm::IsConnOpen(conn)) {
        comm_remove_close_handler(conn->fd, closer);
        closer = nullptr;
        conn->close();
        conn = nullptr;
        return;
    }

    closer->cancel("no longer needed by tunnel");
    closer = nullptr;
    conn = nullptr;
    return;
}

Must(!conn); // connection requires a closer
```

Finally, let's add comments that document the two (non-obvious) states (these comments are already present in this PR) and the third state (the PR should get the same `we do not own a connection` comment):

```C++
if (closer) {
    if (Comm::IsConnOpen(conn)) {
        // We did not call conn->close().
        comm_remove_close_handler(conn->fd, closer);
        closer = nullptr;
        conn->close();
        conn = nullptr;
        return;
    }

    // We called conn->close() followed by this closeQuietly() call.
    // Our close callback has already been scheduled but not yet fired.
    closer->cancel("no longer needed by tunnel");
    closer = nullptr;

    // fd_table[old conn->fd].closing() ought to be true here,
    // but we cannot check that because conn->close() clears conn->fd

    conn = nullptr;
    return;
}

// we do not own a connection
Must(!conn); // connection requires a closer
```

Now it becomes clear that the fixed version uses nested `if` statements to detect exactly the same two states that the PR code detects without nesting. The fixed version also uses a rather unusual for Squid top-level condition -- `closer`; most of Squid code starts by handling the Comm::IsConnOpen(conn) case at the top level.

IMO, the PR version is (a little) better than the code above because it is a little simpler and follows Squid conventions a little better. Do you insist on going with the above version instead? If yes, why?